libbpf-tools/offcputime.c

```diff
@@ -0,0 +1,350 @@
+			argp_usage(state);
+		}
+		break;
+	case ARGP_KEY_ARG:
+		if (pos_args++) {
+			fprintf(stderr,
+				"Unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		errno = 0;
+		env.duration = strtol(arg, NULL, 10);
+		if (errno || env.duration <= 0) {
+			fprintf(stderr, "Invalid duration (in s): %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+}
+
+
+static void print_map(struct ksyms *ksyms, struct syms_vec *syms_vec,
+		      struct offcputime_bpf *obj)
+{
+	struct key_t lookup_key = {}, next_key;
+	const struct ksym *ksym;
+	const struct syms *syms;
+	const struct sym *sym;
+	int err, i, ifd, sfd;
+	unsigned long *ip;
+	struct val_t val;
+
+	ip = calloc(env.perf_max_stack_depth, sizeof(*ip));
+	if (!ip) {
+		fprintf(stderr, "failed to alloc ip\n");
+		return;
+	}
+
+	ifd = bpf_map__fd(obj->maps.info);
+	sfd = bpf_map__fd(obj->maps.stackmap);
+	while (!bpf_map_get_next_key(ifd, &lookup_key, &next_key)) {
+		err = bpf_map_lookup_elem(ifd, &next_key, &val);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup info: %d\n", err);
+			goto cleanup;
+		}
+		lookup_key = next_key;
+		if (val.delta == 0)
+			continue;
+		if (bpf_map_lookup_elem(sfd, &next_key.kern_stack_id, ip) != 0) {
+			fprintf(stderr, "    [Missed Kernel Stack]\n");
+			goto print_ustack;
+		}
+		for (i = 0; i < env.perf_max_stack_depth && ip[i]; i++) {
+			ksym = ksyms__map_addr(ksyms, ip[i]);
+			printf("    %s\n", ksym ? ksym->name : "Unknown");
+		}
+
+print_ustack:
+		if (next_key.user_stack_id == -1)
+			goto skip_ustack;
+
+		if (bpf_map_lookup_elem(sfd, &next_key.user_stack_id, ip) != 0) {
+			fprintf(stderr, "    [Missed User Stack]\n");
+			continue;
+		}
+
+		syms = syms_vec__get_syms(syms_vec, next_key.tgid);
+		if (!syms) {
+			fprintf(stderr, "failed to get syms\n");
+			goto skip_ustack;
+		}
+		for (i = 0; i < env.perf_max_stack_depth && ip[i]; i++) {
+			sym = syms__map_addr(syms, ip[i]);
+			if (sym)
+				printf("    %s\n", sym->name);
+			else
+				printf("    [unknown]\n");
+		}
+
+skip_ustack:
+		printf("    %-16s %s (%d)\n", "-", val.comm, next_key.pid);
+		printf("        %lld\n\n", val.delta);
+	}
+
+cleanup:
+	free(ip);
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct syms_vec *syms_vec = NULL;
+	struct ksyms *ksyms = NULL;
+	struct offcputime_bpf *obj;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+	if (env.user_threads_only && env.kernel_threads_only) {
+		fprintf(stderr, "user_threads_only, kernel_threads_only "
+			"cann't be used together.\n");
```

Comment:
typo: can't (and don't split strings, please check all the other strings as well)

Reply:
Fixed, thanks!